Add UIDelegate API to observe text edit menu appearance and dismissal

whsieh opened the pull request:
#### 0db83bab641f94f01a0c4a5a0a597687cda5b0d6
<pre>
Add UIDelegate API to observe text edit menu appearance and dismissal
<a href="https://bugs.webkit.org/show_bug.cgi?id=243102">https://bugs.webkit.org/show_bug.cgi?id=243102</a>
rdar://96239605

Reviewed by Aditya Keerthi.

Move the private UI delegate methods introduced in `252742@main` into the public API header,
`WKUIDelegate.h`, and adopt this new API in our own testing infrastructure, to observe edit menu
presentation and dismissal.

* Source/WebKit/UIProcess/API/Cocoa/WKUIDelegate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKUIDelegatePrivate.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView willPresentEditMenuWithAnimator:]):
(-[WKContentView willDismissEditMenuWithAnimator:]):
* Tools/WebKitTestRunner/cocoa/TestRunnerWKWebView.mm:
(-[TestRunnerWKWebView webView:willPresentEditMenuWithAnimator:]):
(-[TestRunnerWKWebView webView:willDismissEditMenuWithAnimator:]):
(-[TestRunnerWKWebView _webView:willPresentEditMenuWithAnimator:]): Deleted.
(-[TestRunnerWKWebView _webView:willDismissEditMenuWithAnimator:]): Deleted.

Canonical link: <a href="https://commits.webkit.org/252974@main">https://commits.webkit.org/252974@main</a>
</pre>

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKUIDelegatePrivate.h:
```diff
@@ -256,9 +256,6 @@ struct UIEdgeInsets;
 - (NSInteger)_webView:(WKWebView *)webView dataOwnerForDropSession:(id <UIDropSession>)session WK_API_AVAILABLE(ios(11.0));
 - (NSInteger)_webView:(WKWebView *)webView dataOwnerForDragSession:(id <UIDragSession>)session WK_API_AVAILABLE(ios(11.0));
-// FIXME (243102): Expose these edit menu delegates as public API.
-- (void)_webView:(WKWebView *)webView willPresentEditMenuWithAnimator:(id<UIEditMenuInteractionAnimating>)animator WK_API_AVAILABLE(ios(WK_IOS_TBA));
-- (void)_webView:(WKWebView *)webView willDismissEditMenuWithAnimator:(id<UIEditMenuInteractionAnimating>)animator WK_API_AVAILABLE(ios(WK_IOS_TBA));
```
Can you confirm there are no other clients of this SPI (other than the tests)?

whsieh replied:
Yes, I'm pretty sure it should be fine since we've never shipped the SPI (added in rdar://97454308) in any public (or even non-public) builds.

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKUIDelegate.h:
```diff
@@ -254,6 +254,22 @@ typedef NS_ENUM(NSInteger, WKDialogResult) {
+ * @abstract Called when the web view is about to present its edit menu.
+ *
+ * @param textField    The web view displaying the menu.
+ * @param animator      Appearance animator. Add animations to this object to run them alongside the appearance transition.
```
Spacing looks a bit off.

whsieh replied:
Good catch, fixed! (And also changed the `textField` above to `webView` — it's a really good thing you found this!)

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKUIDelegate.h:
```diff
@@ -254,6 +254,22 @@ typedef NS_ENUM(NSInteger, WKDialogResult) {
+ * @param webView    The web view displaying the menu.
+ * @param animator      Dismissal animator. Add animations to this object to run them alongside the dismissal transition.
+ */
+- (void)webView:(WKWebView *)webView willDismissEditMenuWithAnimator:(id<UIEditMenuInteractionAnimating>)animator WK_API_AVAILABLE(ios(WK_IOS_TBA));
```
Ditto (:261).

whsieh committed:
Add UIDelegate API to observe text edit menu appearance and dismissal
https://bugs.webkit.org/show_bug.cgi?id=243102
rdar://96239605

Reviewed by Aditya Keerthi.

Move the private UI delegate methods introduced in `252742@main` into the public API header,
`WKUIDelegate.h`, and adopt this new API in our own testing infrastructure, to observe edit menu
presentation and dismissal.

* Source/WebKit/UIProcess/API/Cocoa/WKUIDelegate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKUIDelegatePrivate.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView willPresentEditMenuWithAnimator:]):
(-[WKContentView willDismissEditMenuWithAnimator:]):
* Tools/WebKitTestRunner/cocoa/TestRunnerWKWebView.mm:
(-[TestRunnerWKWebView webView:willPresentEditMenuWithAnimator:]):
(-[TestRunnerWKWebView webView:willDismissEditMenuWithAnimator:]):
(-[TestRunnerWKWebView _webView:willPresentEditMenuWithAnimator:]): Deleted.
(-[TestRunnerWKWebView _webView:willDismissEditMenuWithAnimator:]): Deleted.

Canonical link: https://commits.webkit.org/252974@main
